tilings/algorithms/subclass_verification.py

```diff
@@ -0,0 +1,100 @@
+"""
+The main algorithm for detecting when the underlying ungridded permutations on a tiling
+are contained in one of a given set of subclasses.
+"""
+
+from typing import TYPE_CHECKING, List, Optional, Set, Tuple, cast
+
+from permuta import Perm
+from tilings.algorithms import Factor, GriddedPermsOnTiling
+
+if TYPE_CHECKING:
+    from tilings import Tiling
+
+
+class SubclassVerificationAlgorithm:
+    def __init__(self, tiling: "Tiling", perms_to_check: Set[Perm]) -> None:
+        self.tiling = tiling
+        self.perms_to_check = perms_to_check
+        self._subclasses: Optional[Tuple[Perm, ...]] = None
+
+    def quick_pare(self) -> Set[Perm]:
+        """
+        Quickly eliminates some subclasses that need to be checked before generating
+        gridded perms. Often this is able to eliminate all subclasses.
+        """
+
+        # First pare: The local gridded permutations on a tiling with no requirements
+        # are those contained in some cell basis.
+        if len(self.tiling.requirements) == 0:
+            perms_leftover = set()
+
+            cell_basis = self.tiling.cell_basis()
+            all_bases = set(tuple(obs) for obs, _ in cell_basis.values())
+            max_base_len = {base: max(len(b) for b in base) for base in all_bases}
+            for perm in self.perms_to_check:
+                if not any(
+                    len(perm) < max_base_len[base] or perm.avoids(*base)
+                    for base in all_bases
+                ):
+                    perms_leftover.add(perm)
+            return perms_leftover
+
+        # Second pare: If a tiling has a single list requirement, which contains a
+        # single requirement, which is a point, then the cell basis of that particular
+        # positive cell can be used to rule out subclasses.
+        if (
+            len(self.tiling.requirements) == 1
+            and len(self.tiling.requirements[0]) == 1
+            and len(self.tiling.requirements[0][0]) == 1
+        ):
+            pos_cell = self.tiling.requirements[0][0].pos[0]
+            pos_basis = self.tiling.cell_basis()[pos_cell][0]
+            max_pos_basis_len = max(len(b) for b in pos_basis)
+            perms_leftover = set()
+            for perm in self.perms_to_check:
+                if not (len(perm) < max_pos_basis_len or perm.avoids(*pos_basis)):
+                    perms_leftover.add(perm)
+            return perms_leftover
+
+        return self.perms_to_check
+
+    @property
+    def subclasses(self) -> Tuple[Perm, ...]:
+        if self._subclasses is None:
+            self.compute_subclasses()
+        return cast(Tuple[Perm, ...], self._subclasses)
+
+    def compute_subclasses(self) -> None:
+
+        self._subclasses = tuple()
+
+        # It is a waste of time to check a factorable tiling, since we will check its
+        # children eventually.
+        if Factor(self.tiling).factorable():
```

Comment:
I'm think it would be better to check that in the Factory instead. This way we can still use the algorithm on a factorable tiling if we want